src/metabase/upload.clj

```diff
@@ -33,68 +33,102 @@
-;; ancestor of the types for each value in the column.
+;; Upload value-types form a DAG (directed acyclic graph) where each type can be relaxed into any of its ancestors.
+;; We parse each value in the CSV file to the most-specific possible type for each column.
+
+;; The most-specific possible type for a column is the closest common ancestor of the types for each value in the
+;; column, found by walking through the graph in topological order, following edges from left to right.
+;; Note that this type is not guaranteed to be one of the least common ancestors!
+;;
+;; See [[metabase.util.ordered-hierarchy/first-common-ancestor]] for more details.
+;;
+;; <pre><code>
 ;;
 ;;              text
 ;;               |
 ;;               |
-;;          varchar-255┐
-;;        /     / \    │
-;;       /     /   \   └──────────┬
-;;      /     /     \             │
-;;  boolean float   datetime  offset-datetime
-;;     |     │       │
-;;     │     │       │
-;;     |    int     date
-;;     |   /   \
-;;     |  /     \
-;;     | /       \
-;;     |/         \
-;; boolean-or-int  auto-incrementing-int-pk
-;;
-;; `boolean-or-int` is a special type with two parents, where we parse it as a boolean if the whole
-;; column's values are of that type. additionally a column cannot have a boolean-or-int type, but
-;; a value can. if there is a column with a boolean-or-int value and an integer value, the column will be int
-;; if there is a column with a boolean-or-int value and a boolean value, the column will be boolean
-;; if there is a column with only boolean-or-int values, the column will be parsed as if it were boolean
+;;          varchar-255──────
+;;        /      /  \        \
+;;      /      /      \       \
+;;  boolean  float   datetime  offset-datetime
+;;     |       │        │
+;;     |       │        │
+;;     │ *float-or-int* │
+;;     │       │        │
+;;     │       │        │
+;;     |      int      date
+;;     |   /       \
+;;     |  /         \
+;; *boolean-or-int*  auto-incrementing-int-pk
 ;;
 ;; </code></pre>
+;;
+;; We have a number of special "abstract" nodes in this graph:
+;;
+;; - `*boolean-or-int*` is an ambiguous node, that could either be parsed as a boolean or as an integer.
+;; - `*float-or-int*` is an integer with an explicit decimal point.
+;;
+;; While a `*boolean-or-int*` is a genuinely ambiguous value, `*float-or-int*` exist to power our desired value-type
+;; coercion and column-type promotion behaviour.
+;;
+;; - If we encounter a `*float-or-int*` inside an `int` column, then we can safely coerce it down to an integer.
+;; - If we encounter a `float` (i.e. a non-zero fraction component), then we need to promote the column to a `float.`
+;;
+;; Columns can not have an abstract type, which has no meaning outside of inference and reconciliation.
+;; If we are left with an abstract type after having processed all the values, we need to traverse
+;; further up the graph until we reach a concrete `column-type`.
```

Comment:
This last sentence isn't true in the case of appends, since we can coerce abstract types into existing concrete types.

Reply:
Yeah I was imagining the "left with" as meaning after we'd considered the column type, but much better to explicitly explain both cases.